### PR TITLE
Display the data viewer selected sheet when the selected data frame in the dialog's selector is not the same as that of the data viewer

### LINKED ISSUE #4712: Updating the data frame selector with current data frame
If i have three dataframes open say A, B and C, i thought we had the feature where when you move from one dataframe to the other say A to B, when you open/reopen a dialog you used recently then you should have the dataframe selector having the current dataframe selected in this case B. 

What we have now, sometimes it still is on the old data frame, or it has no data frame selected and still has items on the receivers, and okay could be enabled which could cause a bug.

### LINKED ISSUE #7288: Default selected sheet in dialog is different from currently displayed sheet
**Describe the bug**
With many common data operations that refer to a specific data frame, a different data frame than the currently shown one is selected by default. This can be confusing. 

**To Reproduce**
- Open two sheets (say, A and B)
- Select A to be displayed
- Prepare > Data Frame > Filter
- If A is selected in the filter dialog, cancel. If B is selected in the filter dialog, this is an instance of this issue and you can stop here.
- Select B to be displayed
- Prepare > Data Frame > Filter
- A is selected in the filter dialog

**Expected behaviour**
By default, select the currently shown sheet in the filter dialog.

**Screenshots**
image

**Details:**
 - OS: Ubuntu 18.04 (docker-version)
 - R-Instat Version: Commit [cdddbdf](https://github.com/IDEMSInternational/r-instat-docker/commit/cdddbdf665956dbd88505b8e093b26619b844413) from https://github.com/IDEMSInternational/r-instat-docker

**Additional context**
This issue is common to many dialogs, also e.g. Prepare > Data frame > Sort or Prepare > Data Reshape > Column Summaries.
There might be a rationale for this (e.g. do a variety of different column summaries one after the other on the same data set), but the behavior is unintuitive.

### LINKED ISSUE #7371: Minor bug on the selector data frame control to remember the current data frame
The selector data frame control seems to not remember the current data frame, when you open the a dialogue when there is not dataset loaded in the dataview, and reopen the selector dataframe doesn't select the current dataframe by default or when you have different datasets `data1`, `data2`, and `data3`, with `data1` as current data in the `Data View` then open any dialogue, the selector dataframe for the first time will select `data1` by default then close the dialogue, delete `data1`, the current data may probably be `data2` or `data3`, then reopen the dialogue, the selector will not select the current data.
@rdstern @Patowhiz this is the issue I found with the selector and I'm trying to fix in PR #7370, it is not the same with issue #4712 which probably needs discussion before implementation or not.

### PR DESCRIPTION
**Pull request by Patowhiz**
Replaces #7370
Fixes #7371
Fixes #4712
Fixes #7288

@africanmathsinitiative/developers this is ready for review.

### CODE REVIEW

**Comment by Patowhiz** [comment]
When reviewing this PR as a user. You may get the bug being solved by @N-thony in PR #7355.

**Comment by Patowhiz** [comment]
@rdstern changes made here touch almost all dialogs. And the changes themselves span over 14 files (increasing risk of conflicts). You may want to prioritize reviewing this PR. Thanks.

**Comment by lloyddewit** [comment]
@Ivanluv please can you review? thanks

**Comment by N-thony** [comment]
@Patowhiz I have checked the functionality and it works fine.
Just few suggestions
image
1) Could the selected dataset in DataView window as shown in red on the dialogue be a bit wider?
2) Could we have more description i.e a tooltip as you suggested that will help the user to understand the purpose of that name on Dialogue?
3) Could we find another for the selected dataframe name rather than the current position i.e **Data Frame:    `w_kenya_al`**? It is a bit confusing
This change overwrite the change I made in PR #7370, there for I will close it. Thanks.

**Comment by Patowhiz** [comment]
@rdstern I'm waiting for your response in regards to item 3 raised by @N-thony in the [comment](https://github.com/africanmathsinitiative/R-Instat/pull/7404#issuecomment-1108263952) above.

**Review by rdstern — COMMENTED** [review]
@Patowhiz that's an interesting change.  Here is the example I used:
image

I had thought differently, i.e. in the example above to just put `wkenya_stationinfo` in red?  That's just to alert the user, that it is no longer the currently visible data frame.  If we stick with your system then I would like the full name to be given.  

We should check with others, as I am not a great design person.

**Comment by Patowhiz** [comment]
@rdstern your [way](https://github.com/africanmathsinitiative/R-Instat/pull/7404#pullrequestreview-952573196) is a bit advanced from implementation point of view but simpler to the users. So I'll implement that.

**Comment by Patowhiz** [comment]
@rdstern @N-thony this is now ready for review.

**Comment by Patowhiz** [comment]
@rdstern @N-thony @lloyddewit this is now ready for review.

Would be useful if this could be review while @ChrisMarsh82 is here. Then I can add more functionalities required in the .Net data book structures that are needed by the controls. Thanks.

**Review by rdstern — APPROVED** [review]
Great.  If the code is ok @lloyddewit or @ChrisMarsh82 then good to merge

**Comment by lloyddewit** [comment]
> Great. If the code is ok @lloyddewit or @ChrisMarsh82 then good to merge

I can review/merge this PR

**Comment by lloyddewit** [comment]
@Ivanluv On 25 April, I requested your review, did you get a chance to do this?

File: instat/ucrDataFrame.vb [hunk 1]
```diff
@@ -41,75 +47,153 @@ Public Class ucrDataFrame
...
     'If bOnlyLinkedToPrimaryDataFrames then should the primary data frame itself be displayed
     Private bIncludePrimaryDataFrameAsLinked As Boolean = True
-    ' To prevent circular refreshing of data frame list
+    ''' <summary>
+    ''' Used as a flag to prevent circular refreshing of data frame list
+    ''' </summary>    
     Private bSuppressRefresh As Boolean = False
 
+    'todo. this event just be replaced with control value changed event.
+    'the ControlValueChanged is sufficent enough. So delete?
```
Comment (lloyddewit) [inline instat/ucrDataFrame.vb:56]: Your suggestion sounds like a cleaner solution.
Should we implement in this PR?

**Comment by Ivanluv** [comment]
@lloyddewit Sorry for the late replay .The code looks okay from my end

File: instat/ucrDataFrame.vb [hunk 2]
```diff
@@ -41,75 +47,153 @@ Public Class ucrDataFrame
...
+        cboAvailableDataFrames.Items.Clear()
+        'add new items 
+        If bOnlyLinkedToPrimaryDataFrames Then
+            'todo. GetLinkedToDataFrameNames should also be done through the data book
+            'As of 22/04/022 the data book did not have this feature
+            cboAvailableDataFrames.Items.AddRange(frmMain.clsRLink.GetLinkedToDataFrameNames(strPrimaryDataFrame, bIncludePrimaryDataFrameAsLinked).ToArray)
+        Else
+            For Each dataFrame As clsDataFrame In frmMain.DataBook.DataFrames
+                cboAvailableDataFrames.Items.Add(dataFrame.strName)
+            Next
```
Comment (lloyddewit) [inline instat/ucrDataFrame.vb:113]: would this work?
```suggestion
            cboAvailableDataFrames.Items.AddRange(frmMain.DataBook.DataFrames.ToArray)
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@Patowhiz I liked your coding style, thank you.
Just a couple of small questions

**Comment by lloyddewit** [comment]
@Patowhiz it looks like @ChrisMarsh82 and I both reviewed this PR at the same time. I see that this PR is now merged.
Please could you address the comments in a new PR? 
thanks and sorry for the inconvenience

**Comment by Patowhiz** [comment]
@lloyddewit yes, I'll address the issues in a separate PR.

File: instat/ucrDataFrame.vb (hunk 2 above)
Comment (Patowhiz) [reply instat/ucrDataFrame.vb:113]: @lloyddewit I don't think so. What we want here is to display the names of the data frame which are contained in the `strName` property of the  DataFrame objects.

Comment (lloyddewit) [reply instat/ucrDataFrame.vb:113]: yes, you're correct

File: instat/ucrDataFrame.vb (hunk 1 above)
Comment (Patowhiz) [reply instat/ucrDataFrame.vb:56]: This has now been done in PR #7451.

Comment (lloyddewit) [reply instat/ucrDataFrame.vb:56]: great, thanks

**Pull request merged**